accepted/future-releases/3009-inference-using-bounds/design-document.md

```diff
@@ -0,0 +1,620 @@
+is `B`, allowing the motivating example to compile and run without
+errors.
+
+The picture below shows the details of the updated type inference
+process in the motivating example. The constraints collected during
+the upwards phase (1) contribute the lower bound `C` (2a) as the input
+`P` to [Subtype constraint generation][subtype-constraint-generation],
+and the type parameter bound contributes the upper bound `A<X>` (2b)
+as the input `Q` to the same algorithm.  The newly generated
+constraint (2c) is then added to the set of existing constraints (1).
+Then the solution for the type variable is computed form the updated
+constraint set (3).
+
+```mermaid
+block-beta
+  columns 5
+
+  space:2
+  overallInput["Input"]
+  space:2
+
+  space:5
+
+  block:downwards:5
+    columns 5
+
+    commentDownwardsConstraints(["Downwards constraints:"])
+    space:3
+    commentDownwardsBounds(["Bounds:"])
+
+    constraintsDownwards["∅"]
+    space
+    csstvDownwards["CSSTV"]
+    space
+    boundsDownwards["X extends A&lt;X&gt;"]
+  end
+
+  space:5
+
+  block:upwards:5
+    columns 5
+
+    commentUpwardsConstraints(["Upwards constraints:"])
+    space:3
+    commentUpwardsBounds(["Bounds:"])
+
+    constraintsUpwards["C <# X"]
+    space
+    csstvUpwards["CSSTV"]
+    space
+    boundsUpwards["X extends A&lt;X&gt;"]
+  end
+
+  space:10
+
+  space
+  block:constraintsUpwardsUpdated:1
+    columns 1
+    constraintUpwards2["C <# X"]
+    space
+    constraintUpwards1["B <# X"]
+  end
+  space
+  constraintGeneration["C <# A&lt;X&gt; ==> B <# X"]
+  space
+
+  space:2
+  overallOutput["Output"]
+  space:2
+
+  overallInput -- "{X = _}"--> csstvDownwards
+
+  constraintsDownwards --> csstvDownwards
+  boundsDownwards --> csstvDownwards
+  csstvDownwards -- "{X = _}" --> csstvUpwards
+
+  constraintGeneration -- "(2c) New constraints" --> constraintUpwards1
+  constraintUpwards2 -- "(2a) Contribute\nC <# X to C <# A&lt;X&gt;" --> constraintGeneration
+
+  constraintsUpwards -- "(1) Existing constraints\nfrom the\nupwards phase" --> constraintUpwards2
+  constraintsUpwards --> csstvUpwards
+  boundsUpwards --> csstvUpwards
+  boundsUpwards -- "(2b) Contribute A&lt;X&gt; to C <# A&lt;X&gt;\n" --> constraintGeneration
+  constraintsUpwardsUpdated -- "(3) Proceed with CSSTV using\nthe updated constraint set" --> csstvUpwards
+
+  csstvUpwards -- "{X = B}" --> overallOutput
+```
+
+## More examples enabled by the current proposal
+
+In addition to the motivating examples, both the original and the more
+complex, being handled by the current proposal, some cases of non
+F-bounded type variables are enabled by it as well.
+
+The following example demonstrates a tempting assumption that doesn't
+hold today, but will hold with the proposed update. The developers
+might think that they will get the extracted type value for `Y` when
+they specify a type value for `X`. We remember similar cases occurring
+in practice.
```

Comment:
This might also be a little bit confusing... here's a somewhat reorganized way to say the same thing:
```suggestion
With a class like `A` below, and with this feature, the constructor
invocation in `test` is inferred as `A<List<num>, num>(<num>[])`.
Without this feature, it is inferred as `A<List<num>, dynamic>(<num>[])`.

Intuitively, this happens because the type inference algorithm without this
feature is unable to detect that there is a relationship between `X` and `Y`.
With this feature one could say that the value of `Y` is "extracted" from
the value of `X`. We have seen cases in practice demonstrating that the
ability to transfer this information during type inference is useful.
```

Reply:
Thanks for the suggested rewrite! I like it much more than the previous version. Done.